arris-ray/php-src
# Add notifications for PHP 8 backward incompatible changes.

# Overview
This PR adds diagnostic messages that report when some PHP 8 backward incompatible changes will take effect in PHP 7.

A subset of PHP 8 backward incompatible changes already had deprecation or warning notices instrumented in PHP 7. This changeset augments that collection for some conditions that are particularly difficult to identify through linting, static analysis, or manual code inspection.

- Non-strict comparisons between numbers and non-numeric strings will return a different result [[1](https://github.com/arris-ray/php-src/pull/3/files#diff-031b09e4bffeba34f6eb29ac63517dcf7dfc30afc38714296a27139d7e7e4487R2350)]
- Ambiguous, non-absolute trait method references will generate a fatal error [[1](https://github.com/arris-ray/php-src/pull/3/files#diff-16d00017a3ed574cca152e8754246a1424db74e221934eb1c0c2513442882615R1908-R1909)]
- `call_user_func_array` arguments' array keys will be treated as named parameters [[1](https://github.com/arris-ray/php-src/pull/3/files#diff-85701127596aca0e597bd7961b5d59cdde4f6bb3e2a109a22be859ab7568b4d2R3477)]
- The needle argument of various string manipulation functions will permit empty values [[1](https://github.com/arris-ray/php-src/pull/3/files#diff-fcf8a2a38ee4a0e3e2cb7c47251c9920ba8c5886d85969f676f9ddbee7aba503R2032-R2034), [2](https://github.com/arris-ray/php-src/pull/3/files#diff-fcf8a2a38ee4a0e3e2cb7c47251c9920ba8c5886d85969f676f9ddbee7aba503R2093-R2095), [3](https://github.com/arris-ray/php-src/pull/3/files#diff-fcf8a2a38ee4a0e3e2cb7c47251c9920ba8c5886d85969f676f9ddbee7aba503R2181-R2183), [4](https://github.com/arris-ray/php-src/pull/3/files#diff-fcf8a2a38ee4a0e3e2cb7c47251c9920ba8c5886d85969f676f9ddbee7aba503R2276-R2278)]
- The length argument of some string manipulation functions will permit `null` values [[1](https://github.com/arris-ray/php-src/pull/3/files#diff-fcf8a2a38ee4a0e3e2cb7c47251c9920ba8c5886d85969f676f9ddbee7aba503R2414-R2416), [2](https://github.com/arris-ray/php-src/pull/3/files#diff-fcf8a2a38ee4a0e3e2cb7c47251c9920ba8c5886d85969f676f9ddbee7aba503R5814-R5816)]
- Some string manipulation functions will clamp out-of-bound offset arguments instead of returning `false` [[1](https://github.com/arris-ray/php-src/pull/3/files#diff-fcf8a2a38ee4a0e3e2cb7c47251c9920ba8c5886d85969f676f9ddbee7aba503R2419)]
- The argument for string-formatting functions will throw a `TypeError` for non-array types [[1](https://github.com/arris-ray/php-src/pull/3/files#diff-c1a9e60fdf787cbc65c7a93bc7f4e390c7c7333efd2b835f9a968b0b414db0cfR649)]
- `parse_url` will differentiate between empty and missing fragment and query tokens [[1](https://github.com/arris-ray/php-src/pull/3/files#diff-423c77149d4fb8bc54e51d748e1c25906c4207142bdef8043850d4959b43f4a1R310-R311), [2](https://github.com/arris-ray/php-src/pull/3/files#diff-423c77149d4fb8bc54e51d748e1c25906c4207142bdef8043850d4959b43f4a1R322-R323)]
- Math functions will respect the `strict_types` directive [[1](https://github.com/arris-ray/php-src/pull/3/files#diff-8e51ab7f9c143d49351a9e71d73cdab8b84ef38d3a456194c2ee741d8c39703dR485)]
- Invalid escape sequences will no longer be interpreted as literals in regex patterns [[1](https://github.com/arris-ray/php-src/pull/3/files#diff-ceb411098d4cf61761c03f3f729d5c8d1fe585be9099adecd4b8f13d16402fcfR826)]
- `is_callable()` will fail when checking for a non-static method with a classname [[1](https://github.com/arris-ray/php-src/pull/3/files#diff-210c47b94a106559d6926f51c60d8c0d39f996738af3fe79e983a409d87a6b45R3224)]

# QA
- Run `make test TESTS=tests/php8` on the `PHP-7.4.24+diagnostics` branch
- Run `ZEND_DONT_UNLOAD_MODULES=1 USE_ZEND_ALLOC=0 valgrind --leak-check=full --show-reachable=yes --track-origins=yes make test TESTS=tests/ > valgrind.dump 2>&1` on the `PHP-7.4.24` and `PHP-7.4.24+diagnostics` branches 
  - See Valgrind reports in [this Github Gist](https://gist.github.com/arris-ray/10719a43723ff00bf17a1d2b9fdb31d7)
